presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/TestPhoenixDistributedQueries.java

```diff
@@ -0,0 +1,135 @@
+        assertUpdate("INSERT INTO test_delete_compound_pk VALUES (3, 'pkVal3', 'val3')", 1);
+
+        assertUpdate("DELETE FROM test_delete_compound_pk where val1 = 'val2'", 1);
+        assertQuery("SELECT count(*) FROM test_delete_compound_pk where val1 = 'val2'", "SELECT 0");
+
+        assertUpdate("DELETE FROM test_delete_compound_pk where pk1 = 3 AND pk2 = 'pkVal3'", 1);
+        assertQuery("SELECT pk1 from test_delete_compound_pk", "SELECT 1");
+    }
+
+    @Test
+    public void testInsert()
+    {
+        @Language("SQL")
+        String query = "SELECT orderdate, orderkey, totalprice FROM orders";
+
+        assertUpdate("CREATE TABLE test_insert WITH (ROWKEYS='orderkey')AS " + query + " WITH NO DATA", 0);
+        assertQuery("SELECT count(*) FROM test_insert", "SELECT 0");
+
+        assertUpdate("INSERT INTO test_insert " + query, "SELECT count(*) FROM orders");
+
+        assertQuery("SELECT * FROM test_insert", query);
+
+        assertUpdate("INSERT INTO test_insert (orderkey) VALUES (-1)", 1);
+        assertUpdate("INSERT INTO test_insert (orderkey) VALUES (-1)", 1); // Phoenix Upsert
+        assertUpdate("INSERT INTO test_insert (orderkey) VALUES (-2)", 1);
+        assertUpdate("INSERT INTO test_insert (orderkey, orderdate) VALUES (-3, DATE '2001-01-01')", 1);
+        assertUpdate("INSERT INTO test_insert (orderkey, orderdate) VALUES (-4, DATE '2001-01-02')", 1);
+        assertUpdate("INSERT INTO test_insert (orderdate, orderkey) VALUES (DATE '2001-01-03', -5)", 1);
+        assertUpdate("INSERT INTO test_insert (orderkey, totalprice) VALUES (-6, 1234)", 1);
+
+        assertQuery("SELECT * FROM test_insert", query
+                + " UNION ALL SELECT null, -1, null"
+                + " UNION ALL SELECT null, -2, null"
+                + " UNION ALL SELECT DATE '2001-01-01', -3, null"
+                + " UNION ALL SELECT DATE '2001-01-02', -4, null"
+                + " UNION ALL SELECT DATE '2001-01-03', -5, null"
+                + " UNION ALL SELECT null, -6, 1234");
+
+        // UNION query produces columns in the opposite order
+        // of how they are declared in the table schema
+        assertUpdate(
+                "INSERT INTO test_insert (orderkey, orderdate, totalprice) " +
+                        "SELECT orderkey, orderdate, totalprice FROM orders " +
+                        "UNION ALL " +
+                        "SELECT orderkey, orderdate, totalprice FROM orders",
+                "SELECT 2 * count(*) FROM orders");
+
+        assertUpdate("DROP TABLE test_insert");
+
+        assertUpdate("CREATE TABLE test_insert (pk BIGINT WITH (primary_key=true), a ARRAY<DOUBLE>,  b ARRAY<BIGINT>)");
+
+        assertUpdate("INSERT INTO test_insert (pk, a) VALUES (1, ARRAY[null])", 1);
+        assertUpdate("INSERT INTO test_insert (pk, a) VALUES (2, ARRAY[1234])", 1);
+        // An array of numeric primitive types returns 0 when the value is null.
```

Comment:
it should fail instead

Reply:
Hi @findepi .
There is no "insert" or "update" in Phoenix. Everything is an upsert. So this would write a new version of a "row" with the same key.